Rewrite MuJoCo integration docs (#2574)
newton-physics/newton: docs/integrations/mujoco.rst

MuJoCo Integration
==================

> :class:`~newton.solvers.SolverMuJoCo` wraps `mujoco_warp
> <https://github.com/google-deepmind/mujoco_warp>`_ behind Newton's standard
> solver interface. Newton compatible-release-pins (``~=``) both ``mujoco``
> and ``mujoco-warp`` to keep the two version-aligned; see
> :github:`pyproject.toml` for the current pins.

Reviewer: The doc is currently mujoco_warp-centric; the CPU backend (`use_mujoco_cpu=True`) only surfaces in passing in the runtime-sync section. A skimming reader wouldn't know there's a CPU path.

That said — not sure if we want to keep supporting it long-term, so this is more a question than a recommendation. If we do plan to keep CPU support, the suggestion below adds a brief intro mention; if not, leave the doc as-is and let the kwarg quietly fade.

```suggestion
:class:`~newton.solvers.SolverMuJoCo` wraps `mujoco_warp
<https://github.com/google-deepmind/mujoco_warp>`_ behind Newton's standard
solver interface; constructing it with ``use_mujoco_cpu=True`` swaps the
backend to the MuJoCo C reference implementation, useful for CPU-only
environments and debugging numerical mismatches. Newton
compatible-release-pins (``~=``) both ``mujoco`` and ``mujoco-warp`` to
keep the two version-aligned; see :github:`pyproject.toml` for the
current pins.
```

Author: Going the other direction: dropped the one remaining CPU mention (in the Runtime sync section) so the page consistently describes the `mujoco_warp` path. The `use_mujoco_cpu` kwarg stays available but is no longer surfaced in the docs.

Reviewer: That's a good call, thanks.

Because MuJoCo has its own modelling conventions, many Newton properties
are mapped differently or not at all. The sections below describe which
Newton concepts the solver supports, how each is mapped to MuJoCo, how
state is exchanged at every step, and where each piece of the conversion
lives in the source. MuJoCo-specific behaviour that has no Newton-core
equivalent is exposed through the :ref:`custom-attribute namespace
<mujoco-custom-attributes>`. A :ref:`code pointers <mujoco-code-pointers>`

> section at the bottom collects the most useful anchor points.

Reviewer: Add a note clarifying the `mjModel` / `mjData` naming throughout the page so the doc isn't pinned to one backend's struct names.

```suggestion
section at the bottom collects the most useful anchor points.

.. note::

   References to ``mjModel`` / ``mjData`` throughout this page apply
   to both backends — the CPU MuJoCo types and ``mujoco_warp``'s
   ``mjwModel`` / ``mjwData`` analogues. The active structure depends
   on ``use_mujoco_cpu``.
```

Author: Applied as a `.. note::` but rewrote the body. The original text referenced the CPU-vs-warp split and `mjwModel` / `mjwData`, which doesn't fit now that we've dropped CPU-backend mentions. Replaced with a pointer to MuJoCo's `mjModel` / `mjData` reference pages so readers know what `mjData.contact` etc. mean.

Reviewer: Great.

MuJoCo Integration
==================

:class:`~newton.solvers.SolverMuJoCo` wraps `mujoco_warp
<https://github.com/google-deepmind/mujoco_warp>`_ behind Newton's standard

> solver interface. Newton compatible-release-pins (``~=``) both ``mujoco``

Reviewer: nit: "compatible-release-pins" reads as a compound noun used as a verb.

```suggestion
solver interface. Newton uses compatible-release pins (``~=``) on both ``mujoco``
```

Author: Done.

MuJoCo Integration
==================

:class:`~newton.solvers.SolverMuJoCo` wraps `mujoco_warp
<https://github.com/google-deepmind/mujoco_warp>`_ behind Newton's standard
solver interface. A minimal setup looks like this::

    import newton
    from newton.solvers import SolverMuJoCo

    builder = newton.ModelBuilder()
    SolverMuJoCo.register_custom_attributes(builder)
    builder.add_mjcf("robot.xml")        # or build the model directly
    model = builder.finalize()
    solver = SolverMuJoCo(model)
    # ...then call solver.step(state_in, state_out, control, contacts, dt)

> 

Reviewer: Suggest dropping this in favor of the Custom attributes section, where the same snippet sits next to the explanation of why `register_custom_attributes` must come before asset loading.

```suggestion
:class:`~newton.solvers.SolverMuJoCo` wraps `mujoco_warp
<https://github.com/google-deepmind/mujoco_warp>`_ behind Newton's standard
solver interface.
```

Author: Done — dropped from the intro. Snippet stays in the *Custom attributes* section.

Because MuJoCo has its own modelling conventions, many Newton properties
are mapped differently or not at all. The sections below describe which
Newton concepts the solver supports, how each is mapped to MuJoCo, how
state is exchanged at every step, and where each piece of the conversion
lives in the source. MuJoCo-specific behaviour that has no Newton-core
equivalent is exposed through the :ref:`custom-attribute namespace
<mujoco-custom-attributes>`. A :ref:`code pointers <mujoco-code-pointers>`
section at the bottom collects the most useful anchor points.

Coordinate conventions

> ----------------------

Reviewer: Rename to `Conventions` (or `Convention differences`) — the section already covers body ordering in addition to quaternion order, and is likely to grow.

```suggestion
Conventions
-----------
```

Author: Ended up dropping the section entirely rather than renaming. The body-ordering bullet went per your other comment, and on reflection the quaternion-order bullet is in the same boat: both are transparent during `step()` (indexed gather/scatter, joints re-sorted internally) and only surface when you're already inspecting raw `mj_model` / `mj_data` or a `save_to_mjcf` file. Once both bullets were gone, a heading with no content didn't earn its keep.

**Quaternion order.**
  Newton stores quaternions as ``(x, y, z, w)``; MuJoCo uses ``(w, x, y, z)``.
  The solver converts between the two automatically (via
  ``quat_xyzw_to_wxyz`` / ``quat_wxyz_to_xyzw`` in ``kernels.py``). Be aware
  of this when inspecting raw MuJoCo objects — e.g. through ``save_to_mjcf``
  or the solver's ``mj_model``/``mj_data`` attributes.

**Body ordering.**
  The solver sorts joints in depth-first topological order when building
  MuJoCo's kinematic tree. If the Newton model's joint order differs, a

>   warning is emitted because kinematics may otherwise diverge.

Reviewer: Does this need to be called out in the docs? Push/pull (`convert_mj_coords_to_warp_kernel`, `kernels.py:505`) does indexed gather/scatter either way, so no correctness or perf impact on `step()`. Only user-visible in `save_to_mjcf` / direct `mj_data` introspection — where the runtime warning already explains itself when it fires. Am I missing a case?

Author: You weren't missing a case. `SolverMuJoCo._convert_to_mjc` topologically sorts joints itself before emitting them to MuJoCo (`solver_mujoco.py:4232-4240`), and `ModelBuilder.validate_joint_ordering` is opt-in (not run from `finalize`). Combined with the indexed gather/scatter in push/pull, user-side joint order doesn't reach MuJoCo's kinematic tree at all. Dropped the section.

Side note: given the above, the runtime warning at `solver_mujoco.py:4238` ("may lead to diverging kinematics between MuJoCo and Newton") is overstated — happy to send a follow-up softening or removing it if you agree.

Reviewer: Yeah makes sense to fix the warning.

   :widths: 25 35 40

   * - Newton type
     - MuJoCo equivalent
     - Notes

>    * - ``FREE``

Reviewer: Plain literals here violate AGENTS.md: *"Sphinx cross-refs (`:class:`, `:meth:`) with shortest possible targets. Prefer public API paths."* `newton.JointType` and `newton.GeoType` are both publicly exported, so agents writing Newton code currently have to guess the import path. Use `:attr:` refs:

```suggestion
   * - :attr:`~newton.JointType.FREE`
```

Same pattern for every row in this table and in `Geometry types` (`:attr:`~newton.GeoType.SPHERE``, etc.).

       to ``5 × 5 × 5`` when ``shape_size`` is unset.
   * - :attr:`~newton.GeoType.HFIELD`
     - ``mjGEOM_HFIELD``
     - Heightfield data is normalized to ``[0, 1]``; the geom origin is
       shifted by ``min_z`` so the lowest point is at the correct world
       height.
   * - :attr:`~newton.GeoType.MESH` / :attr:`~newton.GeoType.CONVEX_MESH`
     - ``mjGEOM_MESH``
     - MuJoCo only supports **convex** collision meshes. Non-convex meshes
       are convex-hulled automatically, which changes the collision
       boundary. The mesh source's ``maxhullvert`` is forwarded.
   * - :attr:`~newton.GeoType.CONE`, :attr:`~newton.GeoType.GAUSSIAN`
     - *unsupported*
     - Not present in the MuJoCo geom-type map.

**Sites** (shapes with the ``SITE`` flag) are converted to MuJoCo sites —
non-colliding reference frames used for sensor attachment and spatial
tendon wrap anchors.

Several Newton collision features — for example non-convex trimesh,
SDF-based contacts, and hydroelastic contacts — are not part of the
MuJoCo geometry model. They are only available through Newton's
collision pipeline (see `Collision pipeline`_ below).

Actuators
---------

> Newton's per-DOF ``joint_target_mode`` creates MuJoCo general actuators:

Reviewer: ```suggestion
Newton's per-DOF :attr:`~newton.Model.joint_target_mode` creates MuJoCo general actuators:
```

Author: Applied.

       are active).
   * - ``FIXED``
     - *(no joint)*
     - The child body is nested directly under its parent. A fixed joint
       connecting to the world produces a **mocap** body, driven via
       ``mjData.mocap_pos`` / ``mjData.mocap_quat``.
   * - ``DISTANCE``, ``CABLE``
     - *unsupported*
     - Not forwarded to MuJoCo.

Per-DOF ``joint_velocity_limit`` values have no MuJoCo equivalent and are

> ignored.

Reviewer: Duplicates the `Velocity limits are not forwarded.` caveat at rst:428. Drop.

```suggestion
```

Author: Dropped, thanks for catching it.

handles this inline (:github:`newton/_src/utils/import_mjcf.py`); USD
goes through :class:`~newton.usd.SchemaResolverMjc`
(:github:`newton/_src/usd/schemas.py`).

Unsupported MuJoCo features
---------------------------

The sections above describe what Newton forwards *into* MuJoCo. In the
other direction, MuJoCo has several modeling concepts that are not
imported when loading an MJCF or USD asset into Newton, and that
:class:`~newton.solvers.SolverMuJoCo` does not reconstruct during conversion:

- **Sensors** (``<sensor>`` — force/torque, IMU, gyro, accelerometer,
  rangefinder, touch, camera-based, …). Newton has its own sensor
  pipeline (:doc:`/concepts/sensors`) that is independent of the MuJoCo
  solver.
- **Cameras and lights** declared in MJCF/USD. Newton uses its own viewer
  and lighting pipeline; camera/light primitives in the source asset are
  ignored.
- **Keyframes** (``<keyframe>``) — MuJoCo's saved-state / reset mechanism
  is not imported.
- **Composite and flex** (``<composite>``, ``<flex>``) — MuJoCo's built-in
  deformables and soft bodies. Newton has dedicated solvers for cloth,
  MPM, and FEM; they are not part of the MuJoCo integration.

> - **Skinned meshes** (``<skin>``) — visualisation-only, not imported.

Reviewer: nit: missed in the British→American spelling flip — only "visualisation" remains.

```suggestion
- **Skinned meshes** (``<skin>``) — visualization-only, not imported.
```

Author: Done.

   :widths: 25 25 50

   * - Newton type
     - MuJoCo equivalent
     - Notes
   * - :attr:`~newton.GeoType.SPHERE`
     - ``mjGEOM_SPHERE``
     -
   * - :attr:`~newton.GeoType.CAPSULE`
     - ``mjGEOM_CAPSULE``
     -
   * - :attr:`~newton.GeoType.CYLINDER`
     - ``mjGEOM_CYLINDER``
     -
   * - :attr:`~newton.GeoType.BOX`
     - ``mjGEOM_BOX``
     -
   * - :attr:`~newton.GeoType.ELLIPSOID`
     - ``mjGEOM_ELLIPSOID``
     -
   * - :attr:`~newton.GeoType.PLANE`
     - ``mjGEOM_PLANE``
     - Must be attached to a static body (``body=-1``); attaching to a
       non-static body raises ``ValueError`` at conversion time. Planes
       are infinite for collision in MuJoCo regardless of size; the

>        configured ``shape_size`` only affects rendering, defaulting to

Reviewer: `shape_size` isn't a public attribute. `solver_mujoco.py:4152` aliases `model.shape_scale.numpy()` as a local `shape_size`; the public field is `Model.shape_scale`.

```suggestion
       configured :attr:`~newton.Model.shape_scale` only affects rendering, defaulting to
```

  discarded. See `Kinematic links and fixed roots`_.

**Collision filtering bitmask fallback.**
  Newton's :attr:`~newton.Model.shape_collision_group` (see
  :ref:`Collision Groups`) is translated to MuJoCo's ``contype`` /
  ``conaffinity`` via graph coloring
  (:github:`newton/_src/sim/graph_coloring.py`). Up to 32 colors are
  supported (one per ``contype`` bit). If the filtering graph requires
  more, shapes with color index ≥ 32 fall back to ``contype=1`` /
  ``conaffinity=1`` and silently collide with every other shape,
  bypassing the intended filtering and adding extra contact pairs to
  the broadphase.

.. _mujoco-kinematic-links-and-fixed-roots:

Kinematic links and fixed roots
-------------------------------

Newton only allows ``is_kinematic=True`` on articulation roots, so a
"kinematic link" in this section always means a kinematic root body.
Any descendants of that root can still be dynamic and are converted
normally.

At runtime, :class:`~newton.solvers.SolverMuJoCo` keeps kinematic roots
user-prescribed rather than dynamically integrated:

- When converting MuJoCo state back to Newton, the previous Newton

>   :attr:`newton.State.joint_q` and :attr:`newton.State.joint_qd` values

Reviewer: nit: missing `~` shortener; rest of the file uses `:attr:` `~newton...` form.

```suggestion
  :attr:`~newton.State.joint_q` and :attr:`~newton.State.joint_qd` values
```

  are passed through for kinematic roots instead of being overwritten
  from MuJoCo's integrated ``qpos`` and ``qvel``.
- Applied body wrenches and joint forces targeting kinematic bodies
  are ignored on the MuJoCo side.
- Kinematic bodies still participate in contacts, so they can act as
  moving or fixed obstacles for dynamic bodies.

During Newton-to-MuJoCo conversion (at
:class:`~newton.solvers.SolverMuJoCo` construction), roots are mapped
by joint type:

- **Kinematic roots with non-fixed joints** become ordinary MuJoCo
  joints with the same Newton joint type and DOFs. A very large
  internal armature is assigned to those DOFs so MuJoCo treats them
  as prescribed, effectively infinite-mass coordinates.
- **Roots attached to world with a fixed joint** become MuJoCo mocap
  bodies (whether kinematic or not). MuJoCo has no joint coordinates
  for a fixed root, so Newton drives the pose through
  ``mjData.mocap_pos`` and ``mjData.mocap_quat`` instead.
- **World-attached shapes that are not part of an articulation**
  remain ordinary static MuJoCo geometry rather than mocap bodies.

> If you edit :attr:`newton.Model.joint_X_p` or :attr:`newton.Model.joint_X_c`

Reviewer: nit: missing `~` shortener.

```suggestion
If you edit :attr:`~newton.Model.joint_X_p` or :attr:`~newton.Model.joint_X_c`
```

useful entry points. Symbols with a leading underscore are **internal
entry points** — stable enough to navigate to, but not part of the public
API and subject to change.

- :meth:`~newton.solvers.SolverMuJoCo.register_custom_attributes` —
  authoritative registry of every MuJoCo-specific custom attribute and
  frequency.
- :meth:`~newton.solvers.SolverMuJoCo.step` — per-step integration entry
  point.
- ``SolverMuJoCo._convert_to_mjc`` — Newton ``Model`` (and optional
  ``State``) → MuJoCo ``mjModel`` / ``mjData`` (orchestrator).
- ``SolverMuJoCo._init_pairs`` / ``_init_actuators`` / ``_init_tendons`` —
  category-specific parsers that consume the MuJoCo custom attributes.
- ``SolverMuJoCo._apply_mjc_control``,
  ``SolverMuJoCo._update_mjc_data``, and
  ``SolverMuJoCo._update_newton_state`` — per-step control, data, and
  state sync between Newton and MuJoCo.
- :meth:`~newton.solvers.SolverMuJoCo.update_contacts` — explicit pull
  of MuJoCo's resolved contacts into a Newton ``Contacts`` object
  (default per-step path does not pull contacts back).
- :meth:`~newton.solvers.SolverBase.notify_model_changed` —
  re-synchronize MuJoCo state after editing the Newton ``Model`` (e.g.

>   fixed-root pose changes via ``joint_X_p`` / ``joint_X_c``).

Reviewer: nit: same symbols are `:attr:` refs four lines above (L512); use refs here too for consistency.

```suggestion
  fixed-root pose changes via :attr:`~newton.Model.joint_X_p` / :attr:`~newton.Model.joint_X_c`).
```

     -
   * - ``PLANE``
     - ``mjGEOM_PLANE``
     - Must be attached to the world body. Rendered size defaults to
       ``5 × 5 × 5``.
   * - ``HFIELD``
     - ``mjGEOM_HFIELD``
     - Heightfield data is normalized to ``[0, 1]``; the geom origin is
       shifted by ``min_z`` so the lowest point is at the correct world
       height.
   * - ``MESH`` / ``CONVEX_MESH``
     - ``mjGEOM_MESH``
     - MuJoCo only supports **convex** collision meshes. Non-convex meshes
       are convex-hulled automatically, which changes the collision
       boundary. ``maxhullvert`` is forwarded from the mesh source when set.
   * - ``CONE``, ``GAUSSIAN``
     - *unsupported*
     - Not present in the MuJoCo geom-type map.

**Sites** (shapes with the ``SITE`` flag) are converted to MuJoCo sites —
non-colliding reference frames used for sensor attachment and spatial
tendon wrap anchors.

> SDF-based and hydroelastic collisions are not part of the MuJoCo geometry

Reviewer: Newton's pipeline supports more than just SDF / hydroelastic — non-convex trimesh too. Phrasing the list more openly so it doesn't claim completeness:

```suggestion
Several Newton collision features — for example non-convex trimesh,
SDF-based contacts, and hydroelastic contacts — are not part of the
MuJoCo geometry model. They are only available through Newton's
collision pipeline (see `Collision pipeline`_ below).
```

model; they are only available through Newton's collision pipeline (see
`Collision pipeline`_ below).

Mass and inertia
----------------

Bodies with positive mass are exported with ``explicitinertial=True``,
forwarding mass, centre-of-mass offset (``ipos``), and inertia tensor.
When the inertia tensor is diagonal it is passed as a 3-element vector
(preserving MuJoCo's sameframe optimisation, ``body_simple=1``); otherwise
it is passed as the full 6-element symmetric form.

> 

Reviewer: I don't follow the interplay between zero-mass bodies and the `inertiafromgeom = auto` flag here. Can you clarify what happens for (a) a body with no Newton mass and no child geoms (e.g. a pure sensor frame) vs. (b) a body with no Newton mass but child geoms?

   * - ``NONE``

>      - No actuator created for this DOF.

Reviewer: Fold `EFFORT` into the no-actuator row — it falls through `solver_mujoco.py:4679-4716` to the same outcome as `NONE`.

```suggestion
   * - :attr:`~newton.JointTargetMode.NONE`,
       :attr:`~newton.JointTargetMode.EFFORT`
     - No MuJoCo actuator created.
```

The POSITION / VELOCITY / POSITION_VELOCITY rows above (lines 127, 129, 131) are unchanged context so I can't suggestion them — same `:attr:` substitution applies:
- ``POSITION`` → `:attr:`~newton.JointTargetMode.POSITION``
- ``VELOCITY`` → `:attr:`~newton.JointTargetMode.VELOCITY``
- ``POSITION_VELOCITY`` → `:attr:`~newton.JointTargetMode.POSITION_VELOCITY``

Author: Applied.

> ``joint_effort_limit`` is forwarded as ``actfrcrange`` on the joint

Reviewer: ```suggestion
:attr:`~newton.Model.joint_effort_limit` is forwarded as ``actfrcrange`` on the joint
```

Author: Applied.

(prismatic, revolute, and D6) or as ``forcerange`` on the actuator (ball).

The full MuJoCo general-actuator model (arbitrary gain/bias/dynamics types
and parameters, explicit transmission targets, ctrl/force/act ranges) is

> only reachable through the :ref:`mujoco-custom-attributes` namespace.

Reviewer: The bare `:ref:` would render with the renamed section title; use explicit display text.

```suggestion
only reachable through the ``mujoco`` :ref:`custom-attribute namespace <mujoco-custom-attributes>`.
```

Author: Applied.

Additional actuators declared this way are appended after the joint-target
actuators — see ``SolverMuJoCo._init_actuators``.

Equality constraints
--------------------

> 

Reviewer: Anchor before the table so the `data[...]` references make sense to readers without MuJoCo internals knowledge.

```suggestion

Each row's ``data[...]`` reference below points into MuJoCo's
`equality.data <https://mujoco.readthedocs.io/en/stable/XMLreference.html#equality>`_
array; slot layout depends on the constraint type.

```

Author: Applied.

.. list-table::
   :header-rows: 1
   :widths: 20 25 55

   * - Newton type
     - MuJoCo equivalent
     - Notes
   * - :attr:`~newton.EqType.CONNECT`
     - ``mjEQ_CONNECT``
     - Anchor forwarded in ``data[0:3]``.
   * - :attr:`~newton.EqType.WELD`
     - ``mjEQ_WELD``
     - Anchor forwarded in ``data[0:3]``, relative pose in ``data[3:10]``,
       torque scale in ``data[10]``.
   * - :attr:`~newton.EqType.JOINT`
     - ``mjEQ_JOINT``

>      - Polynomial coefficients forwarded in ``data[0:5]``.
>    * - Mimic
>      - ``mjEQ_JOINT``
>      - ``coef0`` / ``coef1`` mapped to polynomial coefficients. Only
>        :attr:`~newton.JointType.REVOLUTE` and
>        :attr:`~newton.JointType.PRISMATIC` joints are supported.

Reviewer: Link to the public builder for mimic constraints.

```suggestion
   * - Mimic
     - ``mjEQ_JOINT``
     - Added via :meth:`~newton.ModelBuilder.add_constraint_mimic`. Maps
       ``coef0`` / ``coef1`` to polynomial coefficients. Only
       :attr:`~newton.JointType.REVOLUTE` and
       :attr:`~newton.JointType.PRISMATIC` joints are supported.
```

Author: Applied.

  rather than hard errors.
- **Contact pairs** — explicit geom-pair contact overrides are parsed by
  ``SolverMuJoCo._init_pairs``.

Collision filtering
-------------------

Newton uses integer ``shape_collision_group`` labels to control which shapes
can collide. MuJoCo uses ``contype``/``conaffinity`` bitmasks with different
semantics: two geoms collide when
``(contype_A & conaffinity_B) || (contype_B & conaffinity_A)`` is non-zero.

The solver bridges the two systems with **graph coloring**. Shapes that must
*not* collide are assigned the same color; each color maps to one bit in
``contype``. ``conaffinity`` is set to the complement so same-color geoms
never match. Up to 32 colors are supported (one per ``contype`` bit). If the
graph requires more than 32 colors, shapes with color index ≥ 32 fall back
to MuJoCo defaults (``contype=1``, ``conaffinity=1``) and will collide with
all other shapes, silently bypassing the intended filtering.

Non-colliding shapes (no ``COLLIDE_SHAPES`` flag, or
``collision_group == 0``) get ``contype = conaffinity = 0``. Body pairs for
which all shape-shape combinations are filtered are emitted as
``<exclude>`` elements.

> 

Reviewer: Most of this section is impl detail (graph-coloring algorithm, bitmask formula, `contype/conaffinity` zeroing for non-colliders, `<exclude>` emission). The user-facing piece — the silent >32-color fallback — fits better as a `Caveats` bullet (separate suggestion below). Drop the whole section.

@adenzler-nvidia — does the >32-color fallback also have measurable perf impact (excess shapes joining the default-collision bucket)? If yes, worth surfacing in the new caveat.

```suggestion
```

Author: Section dropped. Yes, the >32-color fallback does have measurable perf impact: overflow shapes get MuJoCo defaults (`contype=1`, `conaffinity=1`), which means `(1 << N) & 1 ≠ 0` for any color N>0 and `1 & ~(1<<N) ≠ 0`, so an overflowed shape collides with every other shape — including other overflows of the same nominal color. Surfaced this in the new Caveats bullet alongside the silent-filtering point.

Collision pipeline
------------------

By default :class:`~newton.solvers.SolverMuJoCo` uses MuJoCo's built-in
collision detection (``use_mujoco_contacts=True``). Alternatively, you can
set ``use_mujoco_contacts=False`` and pass contacts computed by Newton's

> own collision pipeline into :meth:`~newton.solvers.SolverMuJoCo.step`.

Reviewer: Make it explicit that this is a constructor kwarg.

```suggestion
:class:`~newton.solvers.SolverMuJoCo` uses MuJoCo's built-in collision
detection by default. Construct it with ``use_mujoco_contacts=False``
to feed contacts computed by Newton's own collision pipeline into
:meth:`~newton.solvers.SolverMuJoCo.step` instead.
```

Author: Applied.

Newton's pipeline supports non-convex meshes, SDF-based contacts, and
hydroelastic contacts, which are not available through MuJoCo's collision
detection.

**Multi-contact CCD.** With ``enable_multiccd=True`` the solver allows
up to four contact points per geom pair instead of one. Pairs where
either geom has ``margin > 0`` still fall back to a single contact

> regardless of the flag.

Reviewer: Same constructor-kwarg phrasing, and clarify that this is MuJoCo's `geom_margin` (not the authored `shape_margin`) and cross-link to the next paragraph that covers the Newton↔MuJoCo margin mapping.

```suggestion
**Multi-contact CCD.** Constructing
:class:`~newton.solvers.SolverMuJoCo` with ``enable_multiccd=True``
allows up to four contact points per geom pair instead of one. Pairs
where either geom has non-zero MuJoCo ``geom_margin`` still fall back
to a single contact regardless of the flag (see `Margin zeroing`_ for
how Newton's :attr:`~newton.Model.shape_margin` is forwarded to it).
```

Author: Applied. Switched `` `Margin zeroing`_ `` to "see *Margin zeroing* below" — the bold-prefix paragraph isn't a real section anchor, so `:ref:` resolves to nothing and Sphinx warns. Italics + "below" gets the same intent across without an explicit target.

**Margin zeroing.** ``mujoco_warp`` rejects non-zero geom margins on
box-box pairs (its default NATIVECCD path) and on any box/mesh pair

> when ``enable_multiccd=True``. To stay compatible the solver zeroes

Reviewer: Disambiguate "the solver" from `mujoco_warp` since both are referred to as solvers.

```suggestion
when ``enable_multiccd=True``. To stay compatible :class:`~newton.solvers.SolverMuJoCo` zeroes
```

Author: Applied.

``geom_margin`` model-wide at compile time whenever a box geom exists,
or whenever ``enable_multiccd=True`` is combined with mesh geoms; geoms
with non-zero authored margins emit a warning when

> ``use_mujoco_contacts=True``. The Newton model's ``shape_margin`` array

Reviewer: ```suggestion
``use_mujoco_contacts=True``. The Newton model's :attr:`~newton.Model.shape_margin` array
```

Author: Applied.

is left untouched, and when ``use_mujoco_contacts=False`` the authored
margins are restored at runtime through ``update_geom_properties_kernel``.

Multi-world support
-------------------

When ``separate_worlds=True`` (the default for GPU mode with multiple
worlds), the solver builds a MuJoCo model from the **first world** only and
replicates it across all worlds via ``mujoco_warp``. This requires all
Newton worlds to be structurally identical (same bodies, joints, and
shapes); the solver validates this at construction and raises

> ``ValueError`` on a mismatch.

Reviewer: Same constructor-kwarg phrasing as the use_mujoco_contacts / enable_multiccd paragraphs, plus disambiguate "the solver" → SolverMuJoCo (mujoco_warp does the replication, the wrapper does the structural-identity check).

```suggestion
Constructing :class:`~newton.solvers.SolverMuJoCo` with
``separate_worlds=True`` (the default for GPU mode with multiple
worlds) builds a MuJoCo model from the **first world** only and
replicates it across all worlds via ``mujoco_warp``. This requires
all Newton worlds to be structurally identical (same bodies, joints,
and shapes); :class:`~newton.solvers.SolverMuJoCo` validates this at
construction and raises ``ValueError`` on a mismatch.
```

Author: Applied.

Bodies, joints, equality constraints, and mimic constraints cannot have
a negative world index — assigning any of them to the global world
raises ``ValueError``. Only shapes may live in the global world (-1);
they are shared across all worlds without replication.

Runtime state synchronisation
-----------------------------

Each call to :meth:`~newton.solvers.SolverMuJoCo.step` goes through the
same three-phase cycle:

1. **Push Newton → MuJoCo.** Joint positions and velocities
   (``state_in.joint_q`` / ``joint_qd``) are copied into MuJoCo's
   ``qpos`` / ``qvel`` with quaternion reordering. Control is applied via
   ``control.joint_target_pos`` / ``joint_target_vel`` (→ MuJoCo
   ``ctrl``), ``control.joint_f`` (→ ``qfrc_applied``), and

>    ``state_in.body_f`` (→ ``xfrc_applied``). Direct MuJoCo actuator

Reviewer: Make it strictly a generalization — drop the field-level routing detail (lives in `solver_mujoco.py:3420`) and trim the gating sentence to a substepping pointer.

```suggestion
1. **Push Newton → MuJoCo.** ``SolverMuJoCo._apply_mjc_control`` and
   ``SolverMuJoCo._update_mjc_data`` transfer the Newton ``State``
   and ``Control`` inputs to MuJoCo's working data. When
   ``use_mujoco_contacts=False``, Newton-side contacts are also
   converted before the integrator runs. The joint-state re-sync
   frequency can be controlled via the ``update_data_interval``
   kwarg for substepping schemes.
```

Author: Applied.

   control is exposed through ``control.mujoco.ctrl``. When
   ``use_mujoco_contacts=False``, Newton-side contacts are converted into
   ``mjData.contact`` before the integrator runs. The state sync
   (``joint_q`` / ``joint_qd`` → ``qpos`` / ``qvel``) is gated by the
   ``update_data_interval`` kwarg, which defaults to ``1`` (sync every
   step); larger values amortise the sync, and ``0`` disables it after
   construction so MuJoCo simulates from its own internal state.
2. **Integrate.** ``mujoco_warp`` (or the CPU MuJoCo backend when
   ``use_mujoco_cpu=True``) steps the MuJoCo model forward by ``dt``.
3. **Pull MuJoCo → Newton.** Integrated ``qpos`` / ``qvel`` are copied
   back into ``state_out.joint_q`` / ``joint_qd``, with quaternion

>    reordering; ``body_q`` / ``body_qd`` are recomputed from the new joint

Reviewer: Generalize the field-level enumeration — auxiliary outputs (body_qdd, body_parent_f, qfrc_actuator) are documented in the code.

```suggestion
3. **Pull MuJoCo → Newton.** ``SolverMuJoCo._update_newton_state``
   populates the output ``State`` from the integrated MuJoCo data.
   Kinematic roots pass through unchanged from ``state_in`` (see
   `Kinematic links and fixed roots`_).
```

Author: Applied.

namespace, call
:meth:`~newton.solvers.SolverMuJoCo.register_custom_attributes` on the
:class:`~newton.ModelBuilder` **before** loading any asset or adding
joints/shapes manually::

    import newton
    from newton.solvers import SolverMuJoCo

    builder = newton.ModelBuilder()
    SolverMuJoCo.register_custom_attributes(builder)
    # ...then import MJCF / USD or add bodies/joints/shapes manually...
    model = builder.finalize()

The authoritative list of registered attributes — names, defaults, dtypes,
MJCF / USD source names, and the category each belongs to — is the body
of :meth:`~newton.solvers.SolverMuJoCo.register_custom_attributes` itself.
See :doc:`/concepts/custom_attributes` for how Newton's custom-attribute
system works in general.

A subset of ``mjc:*`` USD attributes is additionally mapped onto Newton's
built-in properties during USD import by
:class:`~newton.usd.SchemaResolverMjc` (:github:`newton/_src/usd/schemas.py`) —
for example, joint-limit stiffness/damping from ``mjc:solreflimit``, and
torsional/rolling friction from ``mjc:torsionalfriction`` /
``mjc:rollingfriction``. Attributes not handled by the schema resolver

> land in the ``mujoco`` namespace unchanged.

Reviewer: Combine the existing two paragraphs under one section that's honest about covering both surfaces.

```suggestion
.. _mujoco-custom-attributes:

MuJoCo-specific parameters in USD and MJCF
------------------------------------------

MuJoCo has parameters with no counterpart in Newton's core API.
:class:`~newton.ModelBuilder` handles them during MJCF / USD import
via two mechanisms.

**Custom-attribute namespace.** A dedicated ``mujoco`` custom-attribute
namespace (``model.mujoco.<name>``) is populated from MJCF elements
and from attributes in the OpenUSD MuJoCo schema (``mjc:*``). To
enable the namespace, call
:meth:`~newton.solvers.SolverMuJoCo.register_custom_attributes` on the
:class:`~newton.ModelBuilder` **before** adding anything to it::

    import newton
    from newton.solvers import SolverMuJoCo

    builder = newton.ModelBuilder()
    SolverMuJoCo.register_custom_attributes(builder)
    # ...then add anything (e.g. import MJCF / USD, add joints, ...)
    model = builder.finalize()

The authoritative list of registered attributes — names, defaults,
dtypes, MJCF / USD source names, and the category each belongs to —
is the body of
:meth:`~newton.solvers.SolverMuJoCo.register_custom_attributes`
itself. See :doc:`/concepts/custom_attributes` for how Newton's
custom-attribute system works in general.

**Direct mapping to Newton built-ins.** Some MuJoCo-specific
attributes are mapped onto Newton's built-in properties during import
(rather than the ``mujoco`` namespace) — for example, joint-limit
stiffness and damping derived from ``solreflimit``. The MJCF parser
handles this inline (:github:`newton/_src/utils/import_mjcf.py`); USD
goes through :class:`~newton.usd.SchemaResolverMjc`
(:github:`newton/_src/usd/schemas.py`).
```

Author: Applied.

Smaller limitations are documented inline where they are most relevant —
see `Caveats`_ below for ``gap``, collision-radius, convex-hull fallback,
and velocity limits; and the unsupported rows in `Joint types`_ and
`Geometry types`_.

Caveats
-------

**geom_gap is always zero.**
  MuJoCo's ``gap`` parameter controls *inactive* contact generation —
  contacts that are detected but do not produce constraint forces until
  penetration exceeds the gap threshold. Newton does not use this concept:
  when the MuJoCo collision pipeline is active it runs every step, so
  there is no benefit to keeping inactive contacts around. Setting
  ``geom_gap > 0`` would inflate ``geom_margin``, which disables MuJoCo's
  multicontact and degrades contact quality. Therefore
  :class:`~newton.solvers.SolverMuJoCo` always sets ``geom_gap = 0``
  regardless of the Newton
  :attr:`~newton.ModelBuilder.ShapeConfig.gap` value. MJCF/USD ``gap``
  values are still imported into
  :attr:`~newton.ModelBuilder.ShapeConfig.gap` on the Newton model, but
  they are not forwarded to the MuJoCo solver.

**shape_collision_radius is ignored.**
  MuJoCo computes bounding-sphere radii (``geom_rbound``) internally from

>   the geometry definition. Newton's ``shape_collision_radius`` is not

Reviewer: ```suggestion
  the geometry definition. Newton's :attr:`~newton.Model.shape_collision_radius` is not
```

Author: Applied.

  forwarded.

**Non-convex meshes are convex-hulled.**
  MuJoCo only supports convex collision geometry. Non-convex ``MESH``
  shapes are automatically convex-hulled at conversion time, changing the
  effective collision boundary.

**Velocity limits are not forwarded.**

>   Newton's ``joint_velocity_limit`` has no MuJoCo equivalent and is

Reviewer: ```suggestion
  Newton's :attr:`~newton.Model.joint_velocity_limit` has no MuJoCo equivalent and is
```

Author: Applied.

>   ignored.

Reviewer: Two new caveats: collision-filtering bitmask fallback (folded from the dropped Collision filtering section) + a CPU+Newton-contacts gotcha I noticed in the runtime-sync review.

```suggestion
  ignored.

**Collision filtering bitmask fallback.**
  Newton's :attr:`~newton.Model.shape_collision_group` (see
  :ref:`Collision Groups`) is translated to MuJoCo's ``contype`` /
  ``conaffinity`` via graph coloring
  (:github:`newton/_src/sim/graph_coloring.py`). Up to 32 colors are
  supported. If the filtering graph requires more, shapes with color
  index ≥ 32 fall back to ``contype=1`` / ``conaffinity=1`` and
  silently collide with every other default-bucket shape, bypassing
  the intended filtering.

**MuJoCo CPU + Newton contacts.**
  ``use_mujoco_contacts=False`` requires the ``mujoco_warp`` backend
  (the default). Combining ``use_mujoco_cpu=True`` with
  ``use_mujoco_contacts=False`` silently drops Newton-side contacts —
  only MuJoCo's own CPU collision detection runs.
```

Author: First caveat (bitmask fallback) applied with an added sentence on the perf impact (extra broadphase pairs from the everyone-collides-with-everyone fallback). Second caveat (CPU + Newton contacts) skipped — decided to drop all CPU-backend references from the page instead, so this gotcha no longer has a home here.

.. _mujoco-kinematic-links-and-fixed-roots:

> Kinematic links and fixed roots
> -------------------------------

Reviewer: "link" is unusual in Newton (entity is "body"). Anchor `.. _mujoco-kinematic-links-and-fixed-roots:` left as-is to preserve the cross-ref from `concepts/articulations.rst:263`.

```suggestion
Kinematic and fixed roots
-------------------------
```

Author: Pushing back on this one. `ModelBuilder.add_link()` (`newton/_src/sim/builder.py:3533`) is real public API — "Adds a link (rigid body) to the model within an articulation." `articulations.rst` itself uses "base link", "body link positions", and `add_link()` examples. "Link" is the natural term for a body inside an articulation, which is exactly what this section is about (kinematic-root behavior only applies to articulation roots). Keeping the title as "Kinematic links and fixed roots".

Reviewer: Ok - in theory this section is only about roots, but makes sense to keep links for agents/users wondering about kinematic links.

Push, pull, and contact-conversion are implemented by
``SolverMuJoCo._apply_mjc_control``, ``SolverMuJoCo._update_newton_state``,
and :meth:`~newton.solvers.SolverMuJoCo.update_contacts`, using kernels
from ``newton/_src/solvers/mujoco/kernels.py`` — see
`Where the conversion lives`_ for the full anchor list.

Solver options
--------------

MuJoCo solver parameters (``iterations``, ``ls_iterations``, ``solver``,
``integrator``, ``cone``, ``jacobian``, ``tolerance``, ``ls_tolerance``,
``impratio``, ``wind``, ``viscosity``, ``density``, ``magnetic``, and the
CCD / SDF iteration counts) follow a three-level resolution priority:

1. **Constructor argument** passed to :class:`~newton.solvers.SolverMuJoCo`
   — one value, applied to all worlds. The full list of kwargs, their
   types, and their defaults is documented on the class itself.
2. **Custom attribute** (``model.mujoco.<option>``) — supports per-world
   values. Typically populated automatically by USD or MJCF import.
3. **Default** — if neither of the above is set, the MuJoCo default is
   used, with one Newton-opinionated exception: ``integrator`` defaults
   to ``implicitfast`` (MuJoCo's default is ``euler``) for better

>    stability on stiff systems.

Reviewer: The section names every parameter but doesn't link to MuJoCo's docs explaining what they do. An agent told to tune solver parameters has the names but no semantics. Add at the end of the section:

    See MuJoCo's `solver documentation
    <https://mujoco.readthedocs.io/en/stable/computation/index.html>`_ and
    `<option> XML reference
    <https://mujoco.readthedocs.io/en/stable/XMLreference.html#option>`_ for
    what each parameter does and when to tune it.

- **Sensors** (``<sensor>`` — force/torque, IMU, gyro, accelerometer,
  rangefinder, touch, camera-based, …). Newton has its own sensor
  pipeline (:doc:`/concepts/sensors`) that is independent of the MuJoCo
  solver.
- **Cameras and lights** declared in MJCF/USD. Newton uses its own viewer
  and lighting pipeline; camera/light primitives in the source asset are
  ignored.
- **Keyframes** (``<keyframe>``) — MuJoCo's saved-state / reset mechanism
  is not imported.
- **Composite and flex** (``<composite>``, ``<flex>``) — MuJoCo's built-in
  deformables and soft bodies. Newton has dedicated solvers for cloth,
  MPM, and FEM; they are not part of the MuJoCo integration.
- **Skinned meshes** (``<skin>``) — visualisation-only, not imported.
- **User plugins** (``<plugin>``) — MuJoCo's plugin mechanism for custom
  passive forces or dynamics is not supported.
- **User data and arbitrary custom elements** (``<custom>``, ``<numeric>``,
  ``<text>``) — not imported. Newton-specific user data should use the
  Newton custom-attribute system instead.
- **Actuator transmission types beyond joint / tendon / site / body /
  slidercrank** — notably, MuJoCo's ``adhesion`` transmission is not

>   supported.

Reviewer: This bullet enumerates the supported transmissions in prose, which drifts on both sides: `jointinparent` is already missing from the list (it's in `SolverMuJoCo.TrnType` and handled by `parse_trntype` in `solver_mujoco.py:968`), and any future change — upstream adding a new transmission, or Newton gaining `adhesion` support — requires a doc edit to stay accurate.

Suggest dropping the enumeration entirely and pointing to the enum as the single source of truth:

```suggestion
- **Actuator transmissions** — not all MuJoCo transmission types are
  supported. See :class:`~newton.solvers.SolverMuJoCo.TrnType` for the
  full list.
```

`SolverMuJoCo.TrnType` is a nested public `IntEnum` at `newton/_src/solvers/mujoco/solver_mujoco.py:193`. This mirrors how the custom-attributes section defers to `register_custom_attributes` rather than listing attributes in prose.

Author: Done — replaced with the pointer to `SolverMuJoCo.TrnType`, verbatim from your suggestion.

During export, :class:`~newton.solvers.SolverMuJoCo` maps roots according
to their joint type:

- **Kinematic roots with non-fixed joints** are exported as ordinary MuJoCo
  joints with the same Newton joint type and DOFs. The solver assigns a
  very large internal armature to those DOFs so MuJoCo treats them like
  prescribed, effectively infinite-mass coordinates.
- **Roots attached to world with a fixed joint** are exported as MuJoCo
  mocap bodies. This applies to both kinematic and non-kinematic Newton
  roots attached to world by :class:`~newton.JointType.FIXED`. MuJoCo has
  no joint coordinates for a fixed root, so Newton drives the pose through
  ``mjData.mocap_pos`` and ``mjData.mocap_quat`` instead.
- **World-attached shapes that are not part of an articulation** remain
  ordinary static MuJoCo geometry rather than mocap bodies.

If you edit :attr:`newton.Model.joint_X_p` or :attr:`newton.Model.joint_X_c`
for a fixed-root articulation after constructing the solver, call
``solver.notify_model_changed(newton.solvers.SolverNotifyFlags.JOINT_PROPERTIES)``
to synchronise the updated fixed-root poses into MuJoCo.

.. _mujoco-code-pointers:

Where the conversion lives

> --------------------------

Reviewer: `Code pointers` reads better in retrieval / TOC than `Where the conversion lives` — terser, matches the `.. _mujoco-code-pointers:` anchor, and the section also lists `step` / `register_custom_attributes` (not pure conversion).

```suggestion
Code pointers
-------------
```

  passed through for kinematic roots instead of being overwritten from
  MuJoCo's integrated ``qpos`` and ``qvel``.
- Applied body wrenches and joint forces targeting kinematic bodies are
  ignored on the MuJoCo side.
- Kinematic bodies still participate in contacts, so they can act as
  moving or fixed obstacles for dynamic bodies.

During export, :class:`~newton.solvers.SolverMuJoCo` maps roots according
to their joint type:

- **Kinematic roots with non-fixed joints** are exported as ordinary MuJoCo
  joints with the same Newton joint type and DOFs. The solver assigns a
  very large internal armature to those DOFs so MuJoCo treats them like
  prescribed, effectively infinite-mass coordinates.
- **Roots attached to world with a fixed joint** are exported as MuJoCo
  mocap bodies. This applies to both kinematic and non-kinematic Newton
  roots attached to world by :attr:`~newton.JointType.FIXED`. MuJoCo has
  no joint coordinates for a fixed root, so Newton drives the pose through
  ``mjData.mocap_pos`` and ``mjData.mocap_quat`` instead.
- **World-attached shapes that are not part of an articulation** remain

>   ordinary static MuJoCo geometry rather than mocap bodies.

Reviewer: Drop "exporter"/"export" terminology — it's not defined elsewhere and conflates SolverMuJoCo's at-construction conversion with the optional `save_to_mjcf` file export. Reword "are exported as" to "become" throughout the export bullets and tighten the kinematic/non-kinematic clause on the mocap row.

Note: the intro paragraph (lines 442-444) needs the same treatment ("MuJoCo exporter" → drop the term, "kinematic link" → "kinematic body", "exported normally" → "converted normally"), but those lines are unchanged context outside this suggestion's anchor — they have to be edited manually.

```suggestion

- When converting MuJoCo state back to Newton, the previous Newton
  :attr:`newton.State.joint_q` and :attr:`newton.State.joint_qd` values
  are passed through for kinematic roots instead of being overwritten
  from MuJoCo's integrated ``qpos`` and ``qvel``.
- Applied body wrenches and joint forces targeting kinematic bodies
  are ignored on the MuJoCo side.
- Kinematic bodies still participate in contacts, so they can act as
  moving or fixed obstacles for dynamic bodies.

During Newton-to-MuJoCo conversion (at
:class:`~newton.solvers.SolverMuJoCo` construction), roots are mapped
by joint type:

- **Kinematic roots with non-fixed joints** become ordinary MuJoCo
  joints with the same Newton joint type and DOFs. A very large
  internal armature is assigned to those DOFs so MuJoCo treats them
  as prescribed, effectively infinite-mass coordinates.
- **Roots attached to world with a fixed joint** become MuJoCo mocap
  bodies (whether kinematic or not). MuJoCo has no joint coordinates
  for a fixed root, so Newton drives the pose through
  ``mjData.mocap_pos`` and ``mjData.mocap_quat`` instead.
- **World-attached shapes that are not part of an articulation**
  remain ordinary static MuJoCo geometry rather than mocap bodies.
```

Author: Bullet rewrite applied verbatim. Reworded the intro paragraph to drop "MuJoCo exporter" while keeping "kinematic link" (see thread on the section title — `add_link()` exists, "link" is the articulation-context term). Also caught one stray "at export time" in the Unsupported MuJoCo features intro and reworded it to "during conversion".

If you edit :attr:`newton.Model.joint_X_p` or :attr:`newton.Model.joint_X_c`
for a fixed-root articulation after constructing the solver, call
``solver.notify_model_changed(newton.solvers.SolverNotifyFlags.JOINT_PROPERTIES)``

> to synchronise the updated fixed-root poses into MuJoCo.

Reviewer: ```suggestion
for a fixed-root articulation after constructing the solver, call
:meth:`~newton.solvers.SolverBase.notify_model_changed` with the
:attr:`~newton.solvers.SolverNotifyFlags.JOINT_PROPERTIES` flag to
synchronise the updated fixed-root poses into MuJoCo.
```

Author: Applied.
